Implement a DuckDuckGo Lite engine

return42 opened the pull request:
## What does this PR do?

Implements a DuckDuckGo Lite engine

## Why is this change important?

The existing [DuckDuckGo engine](https://github.com/searxng/searxng/blob/master/searx/engines/duckduckgo.py) does not support paging. DuckDuckgo-Lite is much faster, [less verbose](https://news.ycombinator.com/item?id=22077861) and does have a paging option.

## How to test this PR locally?

    !ddg paris

Test date ranges and pages

## Related issues

Closes https://github.com/searxng/searxng/issues/349

dalf commented on searx/engines/duckduckgo_lite.py:
```diff
@@ -0,0 +1,159 @@
+        if tr_content.get('class') == 'result-sponsored':
+            continue
+
+        a_tag = eval_xpath_getindex(tr_title, './/td//a[@class="result-link"]', 0)
```
searching for `Paris` in `tr-TR` language, the engine crash.

Possible fix:
```python
        a_tag = eval_xpath_getindex(tr_title, './/td//a[@class="result-link"]', 0, None)
        if not a_tag:
            continue
```

dalf commented on searx/engines/duckduckgo_lite.py:
```diff
@@ -0,0 +1,159 @@
+            continue
+
+        a_tag = eval_xpath_getindex(tr_title, './/td//a[@class="result-link"]', 0)
+        td_content = eval_xpath_getindex(tr_content, './/td[@class="result-snippet"]', 0)
```
Resiliant version with a default empty string:
```
td_content = eval_xpath_getindex(tr_content, './/td[@class="result-snippet"]', 0, '')
```

dalf commented on searx/engines/duckduckgo_lite.py:
```diff
@@ -0,0 +1,159 @@
+# engine dependent config
+categories = ['general']
+paging = True
+supported_languages_url = 'https://duckduckgo.com/util/u172.js'
```
Perhaps another PR: the last URL is https://duckduckgo.com/util/u588.js 

To get this URL:
* download `https://duckduckgo.com/`
* search for `<script type="text/javascript" src="/util/u([0-9]+).js"></script>`

return42 replied:
see 3298fbc

dalf reviewed (changes requested):
* There are some debug code here and there
* When to use duckduckgo.py ? When to use duckduckgo_lite.py ?

return42 committed:
[mod] engine duckduckgo - use DuckDuckGo-Lite

Implement a scrapper for DuckDuckGo-Lite [1].  The existing DuckDuckGo [2]
engine does not support paging.  DuckDuckgo-Lite is much faster, less verbose
and does have a paging option (reversed engineered from the input form of [1]).

[1] https://lite.duckduckgo.com/lite
[2] https://duckduckgo.com/

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] engine duckduckgo - update supported_languages_url

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/engines/duckduckgo.py:
```diff
@@ -63,66 +63,108 @@ def get_region_code(lang, lang_list=None):
     params['data']['q'] = query
-    params['data']['b'] = ''
+
+    # The API is not documented, so we do some reverse engineering and emulate
+    # what https://lite.duckduckgo.com/lite/ does when you press "next Page"
+    # link again and again ..
+
+    params['headers']['Content-Type'] = 'application/x-www-form-urlencoded'
+
+    # initial page does not have an offset
+    if params['pageno'] == 2:
+        # second page does have an offset of 30
+        offset = (params['pageno'] - 1) * 30
+        params['data']['s'] = offset
+        params['data']['dc'] = offset + 1
+
+    elif params['pageno'] > 2:
+        # third and following pages do have an offset of 30 + n*50
+        offset = 30 + (params['pageno'] - 2) * 50
+        params['data']['s'] = offset
+        params['data']['dc'] = offset + 1
+
+    # initial page does not have additional data in the input form
+    if params['pageno'] > 1:
+        # request the second page (and more pages) needs 'o' and 'api' arguments
+        params['data']['o'] = 'json'
+        params['data']['api'] = 'd.js'
+
+    # initial page does not have additional data in the input form
+    if params['pageno'] > 2:
+        # request the third page (and more pages) some more arguments
+        params['data']['nextParams'] = ''
+        params['data']['v'] = ''
+        params['data']['vqd'] = ''
```
Today I improved the paging support, I try to emulate what the input forms from https://lite.duckduckgo.com/lite/  do when you click on "next page" .. while looking along the implementation the "API" seems a bit strange, but this is what the input forms do ...

return42 committed:
[upd] make data.languages

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

unixfox commented on searx/engines/duckduckgo.py:
```diff
@@ -34,21 +42,16 @@
 # search-url
-url = 'https://html.duckduckgo.com/html'
+url = 'https://lite.duckduckgo.com/lite'
 url_ping = 'https://duckduckgo.com/t/sl_h'
```
it's https://duckduckgo.com/t/sl_l instead
